### ISSUE TITLE
Coverall.io integration with MOxUnit tests

### ISSUE THREAD

**Issue by mathieuboudreau** [neuropoly/qMRLab#209]
~**branch**: coverage~
**branch**: [coverage-clean](https://github.com/neuropoly/qMRLab/tree/coverage-clean) (updated) 

After a quite a few debugging detours, I was finally able to integrate [Coveralls.io](https://coveralls.io) into qMRLab with our MOxUnit tests in Octave/Travis by using [MOcov](https://github.com/MOcov/MOcov), which informs readers of what % of the codebase is tested with our MOxUnit tests. A badge with coverage % was added to the README.md file in the [coverage](https://github.com/neuropoly/qMRLab/tree/coverage) branch. 

![coverage](https://user-images.githubusercontent.com/1421029/39735277-794bf6f8-5255-11e8-8a4a-841ab0d7e2e7.png)

It currently sits very low (7%), because right now it only covers the BatchExample_Test (and one model within it); I set it up this way because it was 1 - faster to debug, and 2 - didn't go beyond the 30 minute TRAVIS limit.

There's a few more hurdles I think we'd need to jump (e.h. how to consolidate the fact that we split our tests in jobs with coverall.io, the 30 minute limit (coveralls.io seems to make tests slightly longer)), however I think this would be a really valuable feature to have. One particularly fun feature of this is by clicking the coverage badge, you can then see through the directory exactly which lines in which files were hit/missed by the tests.

**Comment by agahkarakuzu** [comment]
@mathieuboudreau really cool feature! Is there a possibility to activate this feature only for PR builds? Now it makes me wonder the actual coverage. 

Would having more parallel jobs in Travis make this faster? The bottleneck was qmt_spgr with 16 minutes, if all jobs exceed 30 mins we may consider splitting more.

**Comment by mathieuboudreau** [comment]
@agahkarakuzu If you have the time this week, let's have a Skype chat sometime so that I can explain some of the bottlenecks I think we might hit and why – I know less about the TRAVIS setup than you do, so it's possible that some things I think might be an issue are things you might know how to easily solve. 

I'm not sure if this can be set for only PR builds - it's simply run whenever TRAVIS tests are run (it's executed in the TRAVIS config) – so if that option is available for TRAVIS then it should be for this. Is there a reason why you'd want to only run this for PRs?

>Now it makes me wonder the actual coverage.

I've been wondering this for a long time – it's why I've been eager to dive into this haha!

**Comment by agahkarakuzu** [comment]
@mathieuboudreau I will try but not sure, it is a busy week. On the other hand, the definition of "units" is a bit tricky in our case. Is the total will always be the number of m files? 

To cover more, we may need to test fittings by altering fitting options I guess?

**Comment by mathieuboudreau** [comment]
@agahkarakuzu no worries! There is no rush for this feature, we can leave this for after ISMRM, or I might explore it a bit further before the next qMRLab meeting.

>On the other hand, the definition of "units" is a bit tricky in our case. Is the total will always be the number of m files?

I'm not sure I understand this question – but we could discuss it when you have the chance to Skype.

Good luck on your comp prep! A+

**Comment by mathieuboudreau** [comment]
Changed branch to [coverage-clean]( https://github.com/neuropoly/qMRLab/tree/coverage-clean) so that it's a bit cleaner and leaner – will delete branch `coverage` soon.

`coverage-clean` now shows the coverage % badge for the BatchExample_Test for all models except qmt_spgr (which was skipped on purpose), which is 25%. I suspect the test coverage is actually a little bit higher than that, as for some reason some lines that I know for a fact are executed aren't detected as being run by MOcov, which is something I'm still debugging – it might have to do with how we setup the Model objects in our tests, or there might be some bugs in MOcov that lead to those misses (or a mix of both).

Running all MOxUnit TRAVIS tests through the coverage analysis on my local computer & Matlab resulted in a 37.3% coverage (again, I think it's actually slightly higher than that, but not substantially). See below for analysis report.

<details>

```

Lines	Executable	Executed	Coverage	File
18264	7816	2912	37.3%	205 files
27	9	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Addons/SimProtocolOpt/SimCRLB.m
51	15	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Addons/SimProtocolOpt/SimFisherMatrix.m
167	63	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Addons/SimProtocolOpt/Sim_Optimize_Protocol_GUI.m
54	30	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Addons/SimRnd/SimRnd.m
58	40	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Addons/SimRnd/SimRndPlotResults.m
220	109	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Addons/SimRnd/Sim_Multi_Voxel_Distribution_GUI.m
263	85	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Addons/SimVary/Sim_Sensitivity_Analysis_GUI.m
140	50	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Addons/SingleVoxel/Sim_Single_Voxel_Curve_GUI.m
149	64	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/AbstractModel.m
20	13	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/BatchExample_heavy.m
50	27	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/CompareFits.m
133	56	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/FitDataPar.m
18	9	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/FitResultsSave_mat.m
46	23	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/GenerateDocumentation.m
12	6	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/GetError.m
194	110	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/Header_Parse.m
15	6	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/LoadDefaultOptions.m
13	3	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/RandPolyhedron/checkInBoundsAndUptade.m
38	11	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/RandPolyhedron/meshgrid_polyhedron.m
4	2	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/cdmfile.m
23	9	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/computeAmp.m
129	83	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/iqmr_checkHeader.m
30	9	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/pulse/ViewPulse.m
14	9	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/qMRlistModel.m
13	7	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/qMRwriteexample.m
19	6	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/roi_stats.m
83	25	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/sim/BlochSol.m
141	75	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/sim/VaryParam.m
126	68	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/sim/VaryParamPar.m
59	33	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/sim/VaryParamReFit.m
61	32	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/sim/VaryParamView.m
83	42	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/sim/VaryRndParam.m
10	5	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/sim/computeR1obs.m
240	93	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/FileBrowser/BrowserSet.m
245	87	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/FileBrowser/MethodBrowser.m
8	7	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/GUIfun/ClearAxes.m
10	8	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/GUIfun/DrawPlot.m
15	8	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/GUIfun/GetCurrent.m
5	3	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/GUIfun/GetMethod.m
42	28	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/GUIfun/GetPlotRange.m
8	7	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/GUIfun/RefreshColorMap.m
2	1	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/GUIfun/SetAppData.m
27	20	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/GUIfun/UpdatePopUp.m
61	38	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/GUIfun/UpdateSlice.m
49	21	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/GenerateButtons.m
44	17	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/LoadImage.m
14	8	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/ProtLoad.m
5	1	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/addfix.m
336	146	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/dicomLoadAllSeries.m
69	29	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/get_arg_names.m
23	7	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/imagesc3D.m
209	62	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models/T1_relaxometry/vfa_t1.m
122	29	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models/UnderDevelopment/CustomExample.m
125	25	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models/UnderDevelopment/mtv.m
23	16	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/CHARMEDfun/discrete_delta_Delta.m
72	38	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/CHARMEDfun/scd_model_Cyl_GPD.m
57	29	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/CHARMEDfun/scd_schemefile_read.m
67	23	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/Diffusion/scd_preproc_getS0_T2.m
32	9	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/Diffusion/scd_scheme_deltaDeltaTE_constraints.m
127	79	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/Diffusion/scd_scheme_display.m
31	15	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/Diffusion/scd_scheme_display_3D_Delta_delta_G.m
53	20	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/Diffusion/scd_schemefile_FSLconvert.m
8	4	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/Diffusion/scd_schemefile_convertFSL.m
86	20	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/Diffusion/scd_schemefile_create.m
19	5	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/Diffusion/scd_schemefile_write.m
91	37	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/IRfun/rdNls.m
107	35	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MTSATfun/MTSAT_OptionsGUI.m
71	31	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MTV/mtv_correct_receive_profile_v2.m
15	8	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MTV/mtv_fit3dpolynomialmodel.m
11	5	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MTV/mtv_fit3dsplinemodel.m
43	20	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MWF/met2_eva/calc_bkgrnd_noise.m
20	11	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MWF/met2_eva/calc_echo_times.m
8	1	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MWF/met2_eva/err_fit.m
25	12	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MWF/met2_eva/get_avgROI_data.m
107	56	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MWF/met2_eva/loadminc.m
856	348	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MWF/met2_eva/multi_comp_fit.m
47	5	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MWF/met2_eva/prepare_NNLS.m
9	1	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MWF/met2_eva/signal.m
522	287	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SIRFSEfun/SIRFSE_OptionsGUI.m
25	14	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SIRFSEfun/SimDemo_SIRFSE.m
33	12	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SIRFSEfun/functions/FSE_seq.m
15	6	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SIRFSEfun/functions/SIRFSE_GetSeq.m
76	32	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SIRFSEfun/functions/SIRFSE_sim.m
28	15	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/SimDemo_SPGR.m
34	24	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/BuildSfTablePar.m
52	26	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/SPGR_norm.m
133	75	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/SPGR_par.m
118	62	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/SPGR_sim.m
32	19	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/bSSFPfun/SimDemo_bSSFP.m
625	336	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/bSSFPfun/bSSFP_OptionsGUI.m
19	11	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/bSSFPfun/functions/bSSFP_GetSeq.m
119	60	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/bSSFPfun/functions/bSSFP_sim.m
1122	587	0	0.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/qMRLab.m
15	8	2	25.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/CHARMEDfun/scd_model_timedependence.m
32	12	4	33.3%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/CHARMEDfun/scd_scheme2bvecsbvals.m
432	224	104	46.4%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/GUI/Custom_OptionsGUI.m
55	32	15	46.9%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/BuildSfTable.m
518	206	100	48.5%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MWF/met2_eva/multi_comp_fit_v2.m
152	69	36	52.2%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/SPGR_fit.m
17	9	5	55.6%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/GetSf.m
18	5	3	60.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/qMRinfo.m
82	52	33	63.5%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/downloadData.m
364	121	78	64.5%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models/Magnetization_transfer/qmt_sirfse.m
381	133	87	65.4%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models/Diffusion/charmed.m
159	55	36	65.5%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MTV/mtv_compute_m0_t1.m
7	3	2	66.7%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/checkanteriorver.m
85	47	33	70.2%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/j_progress.m
66	21	15	71.4%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/sim/Bloch.m
27	7	5	71.4%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MWF/met2_eva/do_NNLS.m
80	44	32	72.7%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/SPGR_Scw_fun.m
442	143	105	73.4%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models/Magnetization_transfer/qmt_spgr.m
76	19	14	73.7%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/sim/addNoise.m
102	50	37	74.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/Diffusion/scd_display_qspacedata3D.m
77	12	9	75.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models/Noise/noise_level.m
50	24	18	75.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/SPGR_R_fun.m
119	80	62	77.5%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/Noise/scd_noise_fit_histo.m
67	9	7	77.8%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/IRfun/fitT1_IR.m
32	9	7	77.8%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MWF/met2_eva/find_cutoff_index.m
30	10	8	80.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/lineshape/superlorentzianLineshape.m
52	15	12	80.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MWF/met2_eva/do_regNNLS.m
99	48	39	81.2%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/Diffusion/scd_display_qspacedata.m
20	11	9	81.8%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/Diffusion/scd_model_likelihood_rician.m
15	11	9	81.8%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/Diffusion/scd_preproc_getS0.m
51	22	18	81.8%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MTSATfun/MTSAT_exec.m
122	17	14	82.4%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models/FieldMaps/b0_dem.m
9	6	5	83.3%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/qMRloadObj.m
13	6	5	83.3%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/compute_w1cw.m
14	6	5	83.3%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/compute_w1rms.m
67	20	17	85.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/IRfun/getNLSStruct.m
27	20	17	85.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SIRFSEfun/functions/SIRFSE_PlotSimCurve.m
137	47	40	85.1%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/CHARMEDfun/scd_model_CHARMED.m
198	61	52	85.2%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/qMRusage.m
157	69	59	85.5%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/FitData.m
31	22	19	86.4%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Addons/SimVary/SimVaryPlot.m
35	22	19	86.4%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/SPGR_PlotSimCurve.m
75	37	32	86.5%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Addons/SimVary/SimVary.m
22	8	7	87.5%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/computeG.m
19	8	7	87.5%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/computeW.m
40	8	7	87.5%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/pulse/fermi_pulse.m
19	8	7	87.5%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/sct_tool_remove_extension.m
96	57	50	87.7%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/SPGR_Y_fun.m
49	33	29	87.9%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/bSSFPfun/functions/bSSFP_PlotSimCurve.m
235	91	80	87.9%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models/T2_relaxometry/mwf.m
61	25	22	88.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SIRFSEfun/functions/SIRFSE_fun.m
108	42	37	88.1%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/CacheSf.m
289	110	97	88.2%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models/Diffusion/noddi.m
312	106	94	88.7%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models/Diffusion/dti.m
71	28	25	89.3%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SIRFSEfun/functions/SIRFSE_fit.m
18	10	9	90.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/CHARMEDfun/computesigmanoise.m
253	64	58	90.6%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models/Magnetization_transfer/qmt_bssfp.m
24	11	10	90.9%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/compute_w1rp.m
138	56	51	91.1%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/IRfun/rdNlsPr.m
214	92	85	92.4%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/GenerateButtonsWithPanels.m
489	199	187	94.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/qMRgenBatch.m
33	17	16	94.1%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/fwhm.m
58	37	35	94.6%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/button2opts.m
106	58	55	94.8%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/SPGR_Srp_fun.m
66	20	19	95.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/pulse/GetPulse.m
211	43	41	95.3%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models/T1_relaxometry/inversion_recovery.m
47	24	23	95.8%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/CHARMEDfun/ConvertSchemeUnits.m
452	223	215	96.4%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/iqmr_header.m
76	28	27	96.4%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/bSSFPfun/functions/bSSFP_fit.m
54	29	28	96.6%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/sct_tools_ls.m
72	48	47	97.9%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/SPGR_SimCurve.m
22	11	11	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/FitResultsSave_nii.m
8	1	1	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/lineshape/gaussianLineshape.m
8	1	1	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/lineshape/lorentzianLineshape.m
7	4	4	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/list_models.m
9	4	4	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/mat2struct.m
17	7	7	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/objProps2struct.m
34	3	3	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/pulse/gausshann_pulse.m
43	8	8	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/pulse/gaussian_pulse.m
25	1	1	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/pulse/hard_pulse.m
8	4	4	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/pulse/sinc.m
39	6	6	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/pulse/sinc_pulse.m
9	3	3	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/pulse/sincgauss_pulse.m
33	3	3	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/pulse/sinchann_pulse.m
51	28	28	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/qMRshowOutput.m
29	6	6	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/sim/BlochNoMT.m
9	4	4	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/sim/computeR1.m
4	1	1	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/addfixparameters.m
20	10	10	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/button_handle2opts.m
7	3	3	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/cell2str_v3.m
6	4	4	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/extractvoxel.m
16	9	9	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/genvarname_v2.m
7	5	5	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/qMRLabVer.m
10	5	5	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Common/tools/struct2mat.m
77	2	2	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models/FieldMaps/b1_dam.m
120	19	19	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models/Magnetization_transfer/mt_sat.m
15	10	10	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/CHARMEDfun/scd_model_GPD_RDr.m
25	11	11	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/DTIfun/scd_model_dti.m
5	3	3	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/Diffusion/scd_scheme_bvalue.m
7	3	3	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MWF/getDecayMatrix.m
11	5	5	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MWF/getT2.m
58	8	8	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/MWF/met2_eva/iterate_NNLS.m
27	10	10	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/NODDIfun/SchemeToProtocolmat.m
23	15	15	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SIRFSEfun/functions/SIRFSE_SimCurve.m
13	8	8	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SIRFSEfun/functions/computeSr.m
15	8	8	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/SPGR_GetSeq.m
17	11	11	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/SPGR_prepare.m
15	7	7	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/computeSf.m
9	4	4	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/SPGRfun/functions/computeWB.m
34	25	25	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/bSSFPfun/functions/bSSFP_SimCurve.m
55	27	27	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/bSSFPfun/functions/bSSFP_fun.m
18	9	9	100.0%	/Users/mathieuboudreau/neuropoly/github/qMRLab/src/Models_Functions/bSSFPfun/functions/bSSFP_prepare.m
18264	7816	2912	37.3%	205 files

```

</details>

**Comment by agahkarakuzu** [comment]
@mathieuboudreau Travis jobs #13, 14 and 15 almost never complete in the first build run. `spgr` and `bssfp` being `13` and `14` register as a distinct stage due to their title. Interestingly they were completing with no issues when all tests were in the same stage.

Not sure of the robustness of stage `Send combined coveralls.io report`, given that stages are sequential.

Do you have any insight on what may be causing this?

**Comment by mathieuboudreau** [comment]
@agahkarakuzu when MoCov coverage analysis was added, the added function calls (it calls a function everytime a line of code is executed to register it in the coverage analysis) lengthened the tests so much that I had to separate SPGR and bssfp tests – if we reincluded them together today, they would fails 100% of the time due to time constraint.

Send combined coveralls.io report needs to be in a separate stage because all the other tests have to have passed for that stage to run. If all tests were in the same stage, a coverage report would be sent even in the case of failed tests due to time limitations, which could be misleading.

That, and Amazon's free AWS that I'm using to store the coverage analysis files for each test has a maximum number of API calls per month, so having the coveralls.io only send a report when all other tests pass is a way to minimize the API calls (though an API call is still sent for all the tests in the first stage).

We could open an issue re: failing travis tests due to time constraints, I see it happening occasionally too but not all the time (I found sometime TRAVIS takes longer to process tests regardless, even in other repos, I think depending on how busy they might be).
